File: streampark-flink/streampark-flink-kubernetes/src/main/java/org/apache/streampark/flink/kubernetes/PodTemplateParser.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.kubernetes;

import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang3.StringUtils;

import org.yaml.snakeyaml.Yaml;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class PodTemplateParser {

    private static final String KEY_HOST_ALIASES = "hostAliases";
    private static final String KEY_HOSTNAMES = "hostnames";
    private static final String KEY_SPEC = "spec";
    private static final String KEY_IP = "ip";

    public static final String POD_TEMPLATE_INIT_CONTENT =
        "apiVersion: v1\n"
            + "kind: Pod\n"
            + "metadata:\n"
            + "  name: pod-template\n";

    private PodTemplateParser() {
    }

    /** Get init content of pod template */
    public static String getInitPodTemplateContent() {
        return POD_TEMPLATE_INIT_CONTENT.concat("spec:\n");
    }

    /**
     * Complementary initialization pod templates
     *
     * @param podTemplateContent original pod template
     * @return complemented pod template
     */
    @SuppressWarnings("unchecked")
    public static String completeInitPodTemplate(String podTemplateContent) {
        if (podTemplateContent == null || podTemplateContent.trim().isEmpty()) {
            return POD_TEMPLATE_INIT_CONTENT;
        }
        Yaml yaml = new Yaml();
        Map<String, Object> root = yaml.load(podTemplateContent);

        Map<String, Object> res = new LinkedHashMap<>();
        res.put("apiVersion", root.getOrDefault("apiVersion", "v1"));
        res.put("kind", root.getOrDefault("kind", "Pod"));
        Object metadata = root.getOrDefault("metadata", defaultMetadata());
        res.put("metadata", metadata);

        if (root.containsKey("spec")) {
            Object spec = root.get("spec");
            if (spec instanceof Map && !((Map<?, ?>) spec).isEmpty()) {
                res.put("spec", spec);
            }
        }
        return yaml.dumpAsMap(res);
    }

    private static Map<String, Object> defaultMetadata() {
        Map<String, Object> metadata = new LinkedHashMap<>();
        metadata.put("name", "pod-template");
        return metadata;
    }

    /**
     * Add or Merge host alias spec into pod template. When parser pod template error, it would return
     * the origin content.
     *
     * @param hosts hosts info [hostname, ip]
     * @param podTemplateContent pod template content
     * @return pod template content
     */
    @SuppressWarnings("unchecked")
    public static String completeHostAliasSpec(Map<String, String> hosts, String podTemplateContent) {
        if (hosts.isEmpty()) {
            return podTemplateContent;
        }
        try {
            String content = completeInitPodTemplate(podTemplateContent);
            List<Map<String, Object>> hostAlias = covertHostsMapToHostAliasNode(hosts);
            Yaml yaml = new Yaml();
            Map<String, Object> root = yaml.load(content);
            if (!root.containsKey("spec")) {
                Map<String, Object> spec = new LinkedHashMap<>();
                spec.put(KEY_HOST_ALIASES, hostAlias);
                root.put(KEY_SPEC, spec);
                return yaml.dumpAsMap(root);
            }
            Map<String, Object> spec = (Map<String, Object>) root.get(KEY_SPEC);
            spec.put(KEY_HOST_ALIASES, hostAlias);
            return yaml.dumpAsMap(root);
        } catch (Throwable e) {
            return podTemplateContent;
        }
    }

    /** convert hosts map to host alias */
    private static List<Map<String, Object>> covertHostsMapToHostAliasNode(Map<String, String> hosts) {
        Map<String, List<String>> ipToHostnames = new LinkedHashMap<>();
        for (Map.Entry<String, String> entry : hosts.entrySet()) {
            String hostname = entry.getKey().trim();
            String ip = entry.getValue().trim();
            ipToHostnames.computeIfAbsent(ip, k -> new ArrayList<>()).add(hostname);
        }
        List<Map<String, Object>> result = new ArrayList<>();
        for (Map.Entry<String, List<String>> entry : ipToHostnames.entrySet()) {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put(KEY_IP, entry.getKey());
            map.put(KEY_HOSTNAMES, new ArrayList<>(entry.getValue()));
            result.add(map);
        }
        return result;
    }

    /**
     * Extract host-ip map from pod template. When parser pod template error, it would return empty
     * Map.
     *
     * @param podTemplateContent pod template content
     * @return hostname -> ipv4
     */
    @SuppressWarnings("unchecked")
    public static Map<String, String> extractHostAliasMap(String podTemplateContent) {
        Map<String, String> hosts = new LinkedHashMap<>(0);
        if (podTemplateContent == null || podTemplateContent.isEmpty()) {
            return hosts;
        }
        try {
            Yaml yaml = new Yaml();
            Map<String, Object> root = yaml.load(podTemplateContent);
            if (!root.containsKey(KEY_SPEC)) {
                return hosts;
            }
            Map<String, Object> spec = (Map<String, Object>) root.get(KEY_SPEC);
            if (!spec.containsKey(KEY_HOST_ALIASES)) {
                return hosts;
            }
            List<Map<String, Object>> hostAliases = (List<Map<String, Object>>) spec.get(KEY_HOST_ALIASES);
            if (CollectionUtils.isEmpty(hostAliases)) {
                return hosts;
            }
            for (Map<String, Object> hostAlias : hostAliases) {
                collectHostAliasEntry(hosts, hostAlias);
            }
        } catch (Throwable e) {
            return new LinkedHashMap<>(0);
        }
        return hosts;
    }

    @SuppressWarnings("unchecked")
    private static void collectHostAliasEntry(Map<String, String> hosts, Map<String, Object> hostAlias) {
        if (!hostAlias.containsKey(KEY_IP) || !hostAlias.containsKey(KEY_HOSTNAMES)) {
            return;
        }
        String ip = (String) hostAlias.get(KEY_IP);
        if (StringUtils.isBlank(ip)) {
            return;
        }
        List<String> hostnames = (List<String>) hostAlias.get(KEY_HOSTNAMES);
        if (hostnames == null) {
            return;
        }
        for (String hostname : hostnames) {
            if (StringUtils.isNotBlank(hostname)) {
                hosts.put(hostname, ip);
            }
        }
    }

    /**
     * Preview HostAlias pod template content
     *
     * @param hosts hostname -> ipv4
     * @return pod template content
     */
    public static String previewHostAliasSpec(Map<String, String> hosts) {
        List<Map<String, Object>> hostAlias = covertHostsMapToHostAliasNode(hosts);
        Map<String, Object> root = new LinkedHashMap<>();
        root.put(KEY_HOST_ALIASES, hostAlias);
        return new Yaml().dumpAsMap(root);
    }
}

File: streampark-flink/streampark-flink-kubernetes/src/main/scala/org/apache/streampark/flink/kubernetes/PodTemplateParser.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.kubernetes

import org.apache.streampark.common.util.Implicits._

import org.apache.commons.collections.CollectionUtils
import org.apache.commons.lang3.StringUtils
import org.yaml.snakeyaml.Yaml

import java.util

import scala.util.Try
import scala.util.control.Breaks.{break, breakable}

object PodTemplateParser {

  val POD_TEMPLATE_INIT_CONTENT: String =
    """apiVersion: v1
      |kind: Pod
      |metadata:
      |  name: pod-template
      |""".stripMargin

  /** Get init content of pod template */
  def getInitPodTemplateContent: String =
    PodTemplateParser.POD_TEMPLATE_INIT_CONTENT.concat("spec:\n")

  /**
   * Complementary initialization pod templates
   *
   * @param podTemplateContent
   *   original pod template
   * @return
   *   complemented pod template
   */
  def completeInitPodTemplate(podTemplateContent: String): String = {
    if (podTemplateContent == null || podTemplateContent.trim.isEmpty) {
      return POD_TEMPLATE_INIT_CONTENT
    }
    val yaml = new Yaml
    val root = yaml.load(podTemplateContent).asInstanceOf[JavaMap[String, Any]]

    val res = new util.LinkedHashMap[String, Any] {
      put("apiVersion", root.getOrDefault("apiVersion", "v1"))
      put("kind", root.getOrDefault("kind", "Pod"))
      put(
        "metadata",
        root.getOrDefault(
          "metadata", {
            new util.LinkedHashMap[String, Any] {
              put("name", "pod-template")
            }
          }))
    }

    if (root.containsKey("spec")
      && Try(!root.get("spec").asInstanceOf[JavaMap[String, Any]].isEmpty)
        .getOrElse(false)) {
      res.put("spec", root.get("spec"))
    }
    yaml.dumpAsMap(res)
  }

  /**
   * Add or Merge host alias spec into pod template. When parser pod template error, it would return
   * the origin content.
   *
   * @param hosts
   *   hosts info [hostname, ip]
   * @param podTemplateContent
   *   pod template content
   * @return
   *   pod template content
   */
  def completeHostAliasSpec(hosts: JavaMap[String, String], podTemplateContent: String): String = {
    if (hosts.isEmpty) return podTemplateContent
    try {
      val content = completeInitPodTemplate(podTemplateContent)
      // convert hosts map to host alias
      val hostAlias = covertHostsMapToHostAliasNode(hosts)
      // parse yaml
      val yaml = new Yaml
      val root = yaml.load(content).asInstanceOf[JavaMap[String, Any]]
      // no exist spec
      if (!root.containsKey("spec")) {
        val spec = new util.LinkedHashMap[String, Any]()
        spec.put("hostAliases", hostAlias)
        root.put("spec", spec)
        return yaml.dumpAsMap(root)
      }
      // replace spec.hostAliases
      val spec = root.get("spec").asInstanceOf[JavaMap[String, Any]]
      spec.put("hostAliases", hostAlias)
      yaml.dumpAsMap(root)
    } catch {
      case _: Throwable => podTemplateContent
    }
  }

  /** convert hosts map to host alias */
  private[this] def covertHostsMapToHostAliasNode(
      hosts: JavaMap[String, String]): util.ArrayList[util.LinkedHashMap[String, Any]] =
    new util.ArrayList(
      hosts
        .map(e => e._1.trim -> e._2.trim)
        .groupBy(_._2)
        .mapValues(_.keys)
        .toList
        .map(e => {
          val map = new util.LinkedHashMap[String, Any]()
          map.put("ip", e._1)
          map.put("hostnames", new util.ArrayList(e._2.toList))
          map
        }))

  /**
   * Extract host-ip map from pod template. When parser pod template error, it would return empty
   * Map.
   *
   * @param podTemplateContent
   *   pod template content
   * @return
   *   hostname -> ipv4
   */
  def extractHostAliasMap(podTemplateContent: String): JavaMap[String, String] = {
    val hosts = new util.LinkedHashMap[String, String](0)
    if (podTemplateContent == null || podTemplateContent.isEmpty) {
      return hosts
    }
    try {
      val yaml = new Yaml
      val root =
        yaml.load(podTemplateContent).asInstanceOf[JavaMap[String, Any]]
      if (!root.containsKey("spec")) {
        return hosts
      }
      val spec = root.get("spec").asInstanceOf[JavaMap[String, Any]]
      if (!spec.containsKey("hostAliases")) {
        return hosts
      }
      val hostAliases =
        spec.get("hostAliases").asInstanceOf[JavaList[JavaMap[String, Any]]]
      if (CollectionUtils.isEmpty(hostAliases)) {
        return hosts
      }
      for (hostAlias <- hostAliases) {
        breakable {
          if (!hostAlias.containsKey("ip") && !hostAlias.containsKey("hostnames")) break
          val ip = hostAlias.get("ip").asInstanceOf[String]
          if (StringUtils.isBlank(ip)) break
          val hostnames =
            hostAlias.get("hostnames").asInstanceOf[JavaList[String]]
          hostnames
            .filter(StringUtils.isNotBlank(_))
            .foreach(hosts.put(_, ip))
        }
      }
    } catch {
      case _: Throwable => return new util.LinkedHashMap[String, String](0)
    }
    hosts
  }

  /**
   * Preview HostAlias pod template content
   *
   * @param hosts
   *   hostname -> ipv4
   * @return
   *   pod template content
   */
  def previewHostAliasSpec(hosts: JavaMap[String, String]): String = {
    val hostAlias = covertHostsMapToHostAliasNode(hosts)
    val root = new util.LinkedHashMap[String, Any]()
    root.put("hostAliases", hostAlias)
    val yaml = new Yaml
    yaml.dumpAsMap(root)
  }

}
